p2p/transport/webrtc/udpmux/mux.go

```diff
@@ -0,0 +1,194 @@
+package udpmux
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"strings"
+	"sync"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/pion/ice/v2"
+	"github.com/pion/stun"
+)
+
+const receiveMTU = 1500
+
+var _ ice.UDPMux = &udpMux{}
+
+var log = logging.Logger("webrtc-transport-mux")
+
+type ufragConnKey struct {
+	ufrag  string
+	isIPv6 bool
+}
+
+type udpMux struct {
+	mu                   sync.Mutex
+	wg                   sync.WaitGroup
+	ctx                  context.Context
+	cancel               context.CancelFunc
+	socket               net.PacketConn
+	unknownUfragCallback func(string, net.Addr)
+	ufragMap             map[ufragConnKey]*muxedConnection
+	addrMap              map[string]*muxedConnection
+}
+
+func NewUDPMux(socket net.PacketConn, unknownUfragCallback func(string, net.Addr)) ice.UDPMux {
+	ctx, cancel := context.WithCancel(context.Background())
+	mux := &udpMux{
+		ctx:                  ctx,
+		cancel:               cancel,
+		socket:               socket,
+		unknownUfragCallback: unknownUfragCallback,
+		ufragMap:             make(map[ufragConnKey]*muxedConnection),
+		addrMap:              make(map[string]*muxedConnection),
+	}
+
+	mux.wg.Add(1)
+	go mux.readLoop()
+	return mux
+}
+
+// GetConn implements ice.UDPMux
+// It creates a net.PacketConn for a given ufrag if an existing
+// one cannot be  found.
+func (mux *udpMux) GetConn(ufrag string, isIPv6 bool) (net.PacketConn, error) {
+	return mux.getOrCreateConn(ufrag, isIPv6)
+}
+
+// Close implements ice.UDPMux
+func (mux *udpMux) Close() error {
+	select {
+	case <-mux.ctx.Done():
+		return nil
+	default:
+	}
+	mux.cancel()
+	log.Warn("waiting for close")
+	mux.wg.Wait()
+	return nil
+}
+
+// RemoveConnByUfrag implements ice.UDPMux
+func (mux *udpMux) RemoveConnByUfrag(ufrag string) {
+	mux.mu.Lock()
+	defer mux.mu.Unlock()
+	removedAddresses := []string{}
+	for _, isIPv6 := range []bool{true, false} {
+		key := ufragConnKey{ufrag: ufrag, isIPv6: isIPv6}
+		if conn, ok := mux.ufragMap[key]; ok {
+			_ = conn.closeConnection()
+			removedAddresses = append(removedAddresses, conn.addresses...)
+			delete(mux.ufragMap, key)
+		}
+	}
+
+	for _, addr := range removedAddresses {
+		delete(mux.addrMap, addr)
+	}
+}
+
+func (mux *udpMux) getOrCreateConn(ufrag string, isIPv6 bool) (net.PacketConn, error) {
+	select {
+	case <-mux.ctx.Done():
+		return nil, io.ErrClosedPipe
+	default:
+	}
+	key := ufragConnKey{ufrag: ufrag, isIPv6: isIPv6}
+	mux.mu.Lock()
+	// check if the required connection exists
+	if conn, ok := mux.ufragMap[key]; ok {
+		mux.mu.Unlock()
+		return conn, nil
+	}
+
+	conn := newMuxedConnection(mux, ufrag)
+	mux.ufragMap[key] = conn
+	mux.mu.Unlock()
+	return conn, nil
+}
+
+func (mux *udpMux) writeTo(buf []byte, addr net.Addr) (int, error) {
+	return mux.socket.WriteTo(buf, addr)
+}
+
+func (mux *udpMux) readLoop() {
+	buf := make([]byte, receiveMTU)
```

Comment:
It’s a weird separation to reuse the buffer here, and then make a copy later. It would be cleaner to read into a fresh buffer (from the pool) in this for loop.

Reply:
I don't understand. Currently, we reuse the buffer and use `Write` in the packet queue to push to individual connections. If I use a buffer from the pool, I would either have to copy the packet to each muxed connection and return the buffer, or pass the buffer to each muxed connection without copying and return the buffer on `Read`. Wouldn't that result in a large number of buffers and many allocations required if there is a lot of throughput on the mux?